### PR TITLE
Add signInWithNewPassword GraphQL API

### LINKED ISSUE #14: 첫 로그인 시 계정 비밀번호 변경을 위한 새로운 GraphQL mutation 구현
**이슈 설명:**

계정 보안을 강화하는 일환으로, 모든 계정이 처음으로 로그인 시 비밀번호를 변경하도록 할 필요가 있습니다. 이렇게 하면 디폴트 비밀번호가 첫 로그인 즉시 교체되므로 남용될 위험이 최소화됩니다.

이 요구사항을 구현하기 위해, 새로운 GraphQL mutation인 `sign_in_with_new_password(username: !String, oldPassword: !String, newPassword: !String)`를 추가합니다. 이 mutation은 사용자의 `username`과 `oldPassword`를 사용하여 사용자를 인증하고, 데이터베이스에 저장된 `Account`의 `password` 필드 값을 `newPassword`로 바꾸고 `last_signin_time` 필드를 갱신해야 합니다. `oldPassword`와 `newPassword`는 서로 달라야 합니다.

또한, 계정이 한 번도 로그인한 적이 없는 경우 로그인 시도를 받아들이지 않도록 기존의 `signIn` mutation을 수정해야 합니다.

**구현 전:**

`signIn` mutation은 사용자가 이전에 로그인한 적이 없더라도 로그인 시도를 받아들입니다.

**구현 후:**

1. 첫 사용자가 로그인을 시도합니다.
2. `signIn` mutation은 `last_signin_time`이 `None`인 경우 로그인 시도를 거부합니다.
3. 사용자는 `sign_in_with_new_password`를 사용하여 로그인하고 동시에 비밀번호를 변경합니다.
4. `sign_in_with_new_password` mutation은 사용자를 인증하고, `password`와 `last_signin_time`을 업데이트하고, `AuthPayload`를 리턴합니다.

요청인: @sehkone

### PR DESCRIPTION
**Pull request by henry0715-dev**
Close #14

Changed `signIn` GraphQL API logic.
- Returns `Err` if `last_signin_time` of `account` is `None`.

### CODE REVIEW

File: CHANGELOG.md
Comment (sehkone) [inline CHANGELOG.md:71]: This is not necessary because the method is not for users.

File: src/graphql/account.rs
Comment (sehkone) [inline src/graphql/account.rs:404]: Internal methods do not require the Rust documentation. Explanation suffices, if needed.

File: src/graphql/account.rs
Comment (sehkone) [inline src/graphql/account.rs:292]: This method is open to users. It means the Rust documentation should be fulfilled according to [the convention](https://doc.rust-lang.org/rustdoc/how-to-write-documentation.html).

File: CHANGELOG.md
Comment (henry0715-dev) [reply CHANGELOG.md:71]: I'll delete it.

File: src/graphql/account.rs
Comment (sehkone) [inline src/graphql/account.rs:292]: Since this is a open method to users, we need to fulfill the Rust documentation according to [the convention](https://doc.rust-lang.org/rustdoc/how-to-write-documentation.html)

File: src/graphql/account.rs
Comment (sehkone) [inline src/graphql/account.rs:306]: We usually don't mention returning values in a separate section. You can refer to the standard library to check how to write the Rust documentation.

File: src/graphql/account.rs
Comment (sehkone) [inline src/graphql/account.rs:403]: This method doesn't seem to behave consistently. Its behavior depends on `is_update_password`, which is the only differing argument between the two callers. In other words, this method is not the common part between the two callers. When we extract the common part, it should be identical across all participants.

Comment (sehkone) [reply src/graphql/account.rs:403]: If there were any reason to keep this, it should have at least a different name rather than "common".

Comment (henry0715-dev) [reply src/graphql/account.rs:403]: What do you think of the function name below?
```
sign_in_logic_by_update_password
```

File: src/graphql/account.rs
Comment (henry0715-dev) [reply src/graphql/account.rs:306]: I'll delete it.

File: src/graphql/account.rs
Comment (sehkone) [reply src/graphql/account.rs:403]: It would be better to explicitly outline the process in both methods, `sign_in` and `sign_in_with_new_password`. 

I suggest the following pseudocode:

`sign_in`
    - Check the password
    - Check if there was no previous sign-in
    - Check the IP address
    - Check the number of parallel sessions
    - Sign in

`sign_in_with_new_password`
    - Check the password
    - Check the IP address
    - Check the number of parallel sessions
    - Change the password
    - Sign in

While revealing this skeleton in both methods, we need to extract the common parts as separate methods.

Comment (sehkone) [reply src/graphql/account.rs:403]: I want to point out we need to check the password first. The information about whether the user is signing in for the first time or again, whether the number of sessions exceeds the limit, or whether the IP address is valid, is so confidential that it should not be accessible to anyone who has not provided the correct password. The original code was not properly implemented in this regard.

Comment (sehkone) [reply src/graphql/account.rs:403]: > While revealing this skeleton in both methods, we need to extract the common parts as separate methods.

The common part might involve validating the IP address and the number of sessions.

Comment (henry0715-dev) [reply src/graphql/account.rs:403]: I considered splitting the function because it is responsible for so many functions, but there is a lot of overlap, so we made it into one function.
As you mentioned above, we will refactor the checking part by splitting it into functions and calling them.

File: src/graphql/account.rs
Comment (sehkone) [inline src/graphql/account.rs:745]: Could you explain why you put this?

File: CHANGELOG.md
Comment (sehkone) [inline CHANGELOG.md:71]: This is a new API, so should be included in the section "Added".

Comment (sehkone) [reply CHANGELOG.md:71]: Since the `signIn` has been changed, it should be documented.

File: src/graphql/account.rs
Comment (henry0715-dev) [reply src/graphql/account.rs:745]: This is the part set to satisfy the existing test case.
There is also a part where admin/admin does not affect the part you are currently using when running for the first time.
In the future, I thought of the expected behavior of creating a new account after logging in with the admin account, and requesting password change when attempting to log in with the created account.

File: CHANGELOG.md
Comment (henry0715-dev) [reply CHANGELOG.md:71]: I'll change it.

File: CHANGELOG.md
Comment (sehkone) [inline CHANGELOG.md:72]: I think we need to document how the new logic works, as well as the fact that it has changed, because in this case, the new behavior is the primary concern for users.

File: src/graphql/account.rs
Comment (sehkone) [reply src/graphql/account.rs:741]: I don't quite understand. Could you elaborate on what you answered?

Comment (sehkone) [reply src/graphql/account.rs:741]: This `initial_credential` seems to create the first admin account and immediately set its sign-in time. This makes it pointless to alter `sign_in` to check the last sign-in time. What do you think?

File: src/graphql/account.rs
Comment (sehkone) [reply src/graphql/account.rs]: When responding to a reviewer's request for changes, it's better to do so both when you first review the request and when you've completed the requested changes . If I had to choose one, I would prefer the latter.

File: CHANGELOG.md
Comment (henry0715-dev) [reply CHANGELOG.md:81]: Added content.

File: src/graphql/account.rs
Comment (henry0715-dev) [reply src/graphql/account.rs]: Yes, I see.
The function that left document has been removed, making the long document above completely unnecessary.

File: CHANGELOG.md
Comment (sehkone) [reply CHANGELOG.md:81]: The key point is that this API rejects a user's access if it is their first attempt. Therefore, it's not enough to simply mention checking the `lastSigninTime`. Additionally, the word "logic" seems unnecessary. I suggest:

- Changed the `signIn` to reject the first sign-in attempt unless it is done through the new `signInWithNewPassword` API.

In this case, "Changed" or "Updated" is preferred over "Added" because the behavior of the API has actually changed from the user's perspective. Also, the word "reject" or any similar one should be included to convey the key point.

**Comment by sehkone** [comment]
When reseting the admin's password, we need to reset the `last_signin_time` as well.

File: CHANGELOG.md
Comment (henry0715-dev) [reply CHANGELOG.md:81]: I will delete the explanation below and change it to the sentence you suggested.
**"Apply a requirement for a password change upon the initial `signIn`."**

File: src/graphql/account.rs
Comment (henry0715-dev) [reply src/graphql/account.rs]: This part has been restored to its original state.

File: src/graphql/account.rs
Comment (sehkone) [inline src/graphql/account.rs:318]: We don't need to present the signatures of GraphQL APIs in the code because users can not only infer them on their own but also check them in the GraphQL playground.

File: src/graphql/account.rs
Comment (sehkone) [inline src/graphql/account.rs:310]: I don't think this part is necessary because it seems to repeat what was mentioned right above.

File: src/graphql/account.rs
Comment (sehkone) [inline src/graphql/account.rs:322]: I suggest:

```rust
/// # Errors
///
/// Returns `Err` if the password is invalid, this is the first sign-in attempt, the access doesn't originate from a permitted IP address, or the number of sessions exceeds the maximum limit.
```

The style where the sentence starts with a third-person singular verb is a convention in Rust documentation. Additionally, it would be better to specify exactly the cases where errors occur.

**Comment by sophie-cluml** [comment]
@sehkone I know we are still in pr-review stage for this PR, but I think merging PR would require implement new UX flow and design in the client side. So I would like to suggest in advance, that we might merge this PR after we release the upcoming review-web version.

File: src/graphql/account.rs
Comment (sehkone) [reply src/graphql/account.rs]: If you deleted that part, that’s not what I meant. I just wanted to say that the return values should be mentioned according to the Rust documentation convention, like this:

```rust
///
/// Returns blah blah blah
///
```

- The title, `# Returns`, is not necessary.
- The sentence starts with `Returns`, a third-person singular verb.

File: src/graphql/account.rs
Comment (sehkone) [reply src/graphql/account.rs:310]: The part I think is unnecessary:

```rust
/// If the `signIn` GraphQL API response is the error message "a password change is
/// required to proceed", you can call the `signInWithNewPassword` GraphQL API.
```

**Comment by sehkone** [comment]
> @sehkone I know we are still in pr-review stage for this PR, but I think merging PR would require implement new UX flow and design in the client side. So I would like to suggest in advance, that we might merge this PR after we release the upcoming review-web version.

Yes. I won't merge this until what is required is done.

File: src/graphql/account.rs
Comment (henry0715-dev) [reply src/graphql/account.rs:318]: I'll delete it.

File: src/graphql/account.rs
Comment (henry0715-dev) [reply src/graphql/account.rs:322]: I'll change it.

File: src/graphql/account.rs
Comment (henry0715-dev) [reply src/graphql/account.rs:310]: I'll delete it.

File: src/graphql/account.rs
Comment (henry0715-dev) [reply src/graphql/account.rs]: I'll be careful.

File: src/graphql/account.rs
Comment (sehkone) [inline src/graphql/account.rs:309]: It would be better to insert an empty line between the title and the content.

Comment (henry0715-dev) [reply src/graphql/account.rs:309]: Thank you for checking the mistake.

**Comment by sehkone** [comment]
@msk 

This PR hasn't been merged yet because it requires the corresponding feature to be added to the UI. Without this feature, users essentially cannot sign in the system.

In the meantime, since requiring users to change their password at the first sign-in is still inconvenient, I think it would be better to make this optional based on a command line option or another approach. Could you suggest your idea on this?

**Comment by msk** [comment]
> This PR hasn't been merged yet because it requires the corresponding feature to be added to the UI. Without this feature, users essentially cannot sign in the system.

I acknowledge that the PR cannot be merged until the corresponding UI feature is in place.

> In the meantime, since requiring users to change their password at the first sign-in is still inconvenient, I think it would be better to make this optional based on a command line option or another approach. Could you suggest your idea on this?

I believe it would be more appropriate to discuss potential changes to the original issue (#14) there rather than within the PR itself. I’ll move this part of the discussion to the issue thread.

File: CHANGELOG.md
Comment (sehkone) [inline CHANGELOG.md:13]: @henry0715-dev,

Could you take a look at this? It indicates that unnecessary spaces are included, which suggests you may need to check your editor setup, as VS Code shouldn't have done this if it were set up properly.

Comment (henry0715-dev) [reply CHANGELOG.md]: I've completed the modification.
Thank you.

File: src/graphql/account.rs
Comment (sophie-cluml) [inline src/graphql/account.rs:1747]: Could you please clarify the goals of this section of the test code?

From what I understand, lines 1704-1731 seem to be validating that a password change is required to proceed at first sign-in. However, I'm not entirely certain about the purpose of lines 1733-1747. Specifically, I’m curious about why `update_account_last_signin_time` is called and why an incorrect username is passed to the signIn GraphQL API. Were you aiming to conduct a general test of the `signIn` function with these steps? If so, I'd appreciate if you could come up with more general naming for the test code function. If you were aiming just to make sure that users cannot proceed without the password change, I think we can just get rid of L1733-1747, since I think `password_validate_proceed` test code targets to cover that scenario.

**Review by sophie-cluml — COMMENTED** [review]
I'd like to ask you for a rebase please.

**Comment by sophie-cluml** [comment]
Thank you for your kindness in describing your PR.  I would like to suggest a small update to align with our convention. Please consider changing

> Changed `sign-in` GraphQL API logic.

to

- Changed `signIn` GraphQL API logic.

File: src/graphql/account.rs
Comment (sophie-cluml) [inline src/graphql/account.rs:467]: I would like to suggest using `validate_last_sign_in_time` instead of `validate_last_signin_time` for consistency. 

From what I googled online, I got this search result, so I am leaning to 'sign in', rather than `signin` or `sign-in`.

image

File: CHANGELOG.md
Comment (sophie-cluml) [inline CHANGELOG.md:15]: In line with this [comment](https://github.com/aicers/review-web/pull/291#issuecomment-2449076264), I noticed that the `signIn` API logic has been modified. As this logic impacts end users, I think it seems beneficial to explicitly include this change in the changelog to keep our users informed. I’d suggest adding a line in the Changed section, perhaps like the following:

- The `signIn` GraphQL API now returns an error if the account was never signed in before. This change is part of enhancing account security by requiring all users to update their passwords upon their first signing in.

File: src/graphql/account.rs
Comment (henry0715-dev) [reply src/graphql/account.rs:1747]: I'll remove it.
Thank you.

File: src/graphql/account.rs
Comment (henry0715-dev) [reply src/graphql/account.rs:344]: I'll change it.

**Comment by sehkone** [comment]
> I'd like to ask you for a rebase please.

@sophie-cluml,

I don't see any conflicts with the main branch. Why would you think this needs a rebase?

**Comment by sehkone** [comment]
> Thank you for your kindness in describing your PR. I would like to suggest a small update to align with our convention. Please consider changing
> 
> > Changed `sign-in` GraphQL API logic.
> 
> to
> 
> * Changed `signIn` GraphQL API logic.

The representation, such as `signIn`, is not merely conventional. Since it is defined in the schema, we must use it as is.

**Comment by sehkone** [comment]
> > Thank you for your kindness in describing your PR. I would like to suggest a small update to align with our convention. Please consider changing
> > > Changed `sign-in` GraphQL API logic.
> > 
> > 
> > to
> > 
> > * Changed `signIn` GraphQL API logic.
> 
> The representation, such as `signIn`, is not merely conventional. Since it is defined in the schema, we must use it as is.

One more thing: backticks should be used to quote elements that belong to another context. When we search for a quoted string in the original text, it must appear exactly as quoted. There is no `sign-in` in our codebase. Across the code and the UI, we only have two representations: `sign_in` and `signIn`.

**Comment by henry0715-dev** [comment]
> > I'd like to ask you for a rebase please.
> 
> @sophie-cluml,
> 
> I don't see any conflicts with the main branch. Why would you think this needs a rebase?

When I commented, it was conflicting.
Currently, rebase is applied.

**Comment by sehkone** [comment]
> > > > Thank you for your kindness in describing your PR. I would like to suggest a small update to align with our convention. Please consider changing
> > > > > Changed `sign-in` GraphQL API logic.
> > > > 
> > > > 
> > > > to
> > > > 
> > > > * Changed `signIn` GraphQL API logic.
> > > 
> > > 
> > > The representation, such as `signIn`, is not merely conventional. Since it is defined in the schema, we must use it as is.
> > 
> > 
> > One more thing: backticks should be used to quote elements that belong to another context. When we search for a quoted string in the original text, it must appear exactly as quoted. There is no `sign-in` in our codebase. Across the code and the UI, we only have two representations: `sign_in` and `signIn`.
> 
> Commit message modification completed. Thank you.

For clarity, I need to augment my previous explanation. Even though backticks are for quoting code elements, they can be used for any string that might or will appear in any code context. Therefore, it doesn't mean they implicate that the string enclosed by them must be searchable. But, here in this context, i.e. CHANGELOG.md, it's assumed that it should be searchable.

To clarify further: while backticks are typically used to quote code elements, they can also be applied to any string that might or will appear in a code context. In other words, backticks doesn’t necessarily imply that the enclosed string must be searchable. However, in this CHANGELOG.md, it’s assumed that it should be searchable.

**Comment by henry0715-dev** [comment]
> > > > > Thank you for your kindness in describing your PR. I would like to suggest a small update to align with our convention. Please consider changing
> > > > > > Changed `sign-in` GraphQL API logic.
> > > > > 
> > > > > 
> > > > > to
> > > > > 
> > > > > * Changed `signIn` GraphQL API logic.
> > > > 
> > > > 
> > > > The representation, such as `signIn`, is not merely conventional. Since it is defined in the schema, we must use it as is.
> > > 
> > > 
> > > One more thing: backticks should be used to quote elements that belong to another context. When we search for a quoted string in the original text, it must appear exactly as quoted. There is no `sign-in` in our codebase. Across the code and the UI, we only have two representations: `sign_in` and `signIn`.
> > 
> > 
> > Commit message modification completed. Thank you.
> 
> To clarify further: while backticks are typically used to quote code elements, they can also be applied to any string that might or will appear in a code context. In other words, backticks doesn’t necessarily imply that the enclosed string must be searchable. However, in this CHANGELOG.md, it’s assumed that it should be searchable.

Thanks for the detailed explanation and guide. 
I will be careful with backticks and naming.

File: src/graphql/account.rs
Comment (henry0715-dev) [reply src/graphql/account.rs]: Modification completed.

File: CHANGELOG.md
Comment (sophie-cluml) [inline CHANGELOG.md:13]: I think we need to break the line before the word "signed". Could you check your IDE settings?

Comment (henry0715-dev) [reply CHANGELOG.md]: I checked for modification using the cmd + q command in the VSCode IDE.
thank you.

**Review by sophie-cluml — APPROVED** [review]

**Comment by sophie-cluml** [comment]
Could you rebase please?

**Comment by henry0715-dev** [comment]
> Could you rebase please?
It's done.

**Review by sophie-cluml — APPROVED** [review]

**Pull request merged**
